src/Analyzers/Core/Analyzers/RemoveUnnecessarySuppressions/AbstractRemoveUnnecessaryPragmaSuppressionsDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,558 @@
+            if (excludeAll)
+            {
+                return;
+            }
+
+            var root = tree.GetRoot(cancellationToken);
+
+            // Bail out if there are no pragma directives or tree has syntax errors.
+            if (!root.ContainsDirectives || root.GetDiagnostics().Any(d => d.Severity == DiagnosticSeverity.Error))
+            {
+                return;
+            }
+
+            // Process pragma directives in the tree.
+            // The core algorithm is as follows:
+            //  1. Iterate through all the active pragmas in the source file and identify the pragmas
+            //     with diagnostics IDs for which we support unnecesary pragma analysis.
+            //  2. Build the following data structures during this loop:
+            //      a. A map from diagnostic ID to list of pragmas for the ID. This map tracks supported diagnostic IDs for this tree's pragmas.
+            //      b. A array of tuples of candidate pragmas sorted by span, along with associated IDs and enable/disable flag.
+            //         This sorted array allows mapping an unnecessary pragma to the corresponding toggling pragma pair for removal.
+            //      c. A map from pragmas to a boolean indicating if the pragma was used or not.
+            //      d. A set of supported compiler diagnostic IDs that are used in pragmas in this file.
+            //  3. Map the set of candidate diagnostic IDs to the analyzers that can report diagnostics with these IDs.
+            //  4. Execute these analyzers to compute the diagnostics reported by these analyzers in this file.
+            //  5. Iterate through the suppressed diagnostics from this list, and mark the closest preceeeding disable pragma
+            //     which suppresses this ID as used/necessary. Also mark the matching restore pragma as used.
+            //  6. Finally, report a diagostic all the pragmas which have not been marked as used.
+
+            var hasPragmaInAnalysisSpan = false;
+            using var _1 = PooledDictionary<string, List<(SyntaxTrivia pragma, bool isDisable)>>.GetInstance(out var idToPragmasMap);
+            using var _2 = ArrayBuilder<(SyntaxTrivia pragma, ImmutableArray<string> ids, bool isDisable)>.GetInstance(out var sortedPragmasWithIds);
+            using var _3 = PooledDictionary<SyntaxTrivia, bool>.GetInstance(out var pragmasToIsUsedMap);
+            using var _4 = ArrayBuilder<string>.GetInstance(out var idsBuilder);
+            using var _5 = PooledHashSet<string>.GetInstance(out var compilerDiagnosticIds);
+            foreach (var trivia in root.DescendantTrivia())
+            {
+                // Check if this is an active pragma with at least one applicable diagnostic ID.
+                if (SyntaxFacts.IsPragmaDirective(trivia, out var isDisable, out var isActive, out var idNodes) &&
+                    isActive &&
+                    idNodes.Count > 0)
```

Comment:
idnodes is not very clear to me in terms of what it means.

Reply:
maybe provide example?  esp. how you can have separate nodes.

Reply:
since you don't process inactive pragmas, why not `IsActivePragmaDirective`?

Reply:
`IsPragmaDirective` implementation was already used in suppression code fixer, and that has no active pragma check. So, I kept it generic to allow using the same syntax facts API in both places.

Reply:
Sure, I can add an example here.